### ISSUE TITLE
Try fst to support fuzzy search

### ISSUE THREAD

**Issue by mre** [tinysearch/tinysearch#118]
https://github.com/BurntSushi/fst
https://blog.burntsushi.net/transducers/#common-crawl

Expect that this blows up the index size significantly, but it might be an alternative backend that could be enabled with a feature flag, e.g. `tinysearch --engine=fst` or so.
@BurntSushi, do you think that a corpus like that would compress nicely with fst? https://github.com/tinysearch/tinysearch/blob/master/fixtures/index.json
Think I'll play around with your sample code on the main repository when I find the time.

**Comment by BurntSushi** [comment]
Not sure I know how to answer that without more details on how you're building the index. "fuzzy search" can refer to many different things... I'd be happy to say more if you shared more of your plan. :-)

**Comment by mre** [comment]
Sure, sorry. [Here](https://endler.dev/2019/tinysearch/)'s a writeup. It's basically using a bloom filter now but only full words can be searched. What I like to have is an optional, alternative engine that could provide prefix matches and "typo correction" (e.g. configurable Levenshtein). From what I can tell, all of this is supported by fst and it was recommended to me by two independent people, so I thought about giving it a try. My only concern is that the output size of the final (wasm) package, which contains the index will be too big for serving it directly to the clients.

**Comment by BurntSushi** [comment]
I guess what I meant was, what specifically would be the keys in your FST? I see your JSON file, but there are many different things there. Is it every word in the `body` field? Ever 3-gram in the `body` field? And what are the values that your keys map to?

Either way, it's hard for me to say up front. Just think of an FST as a finite state machine. If there are a lot of common prefixes or suffixes, then an FST should be able to exploit that. I would, however, expect a bloom filter to give you more control over the size of your index. e.g., You can make trade offs like "give me a smaller filter perhaps at the expense of a higher false positive rate" more easily than you can with an FST.

But yes, if your FST keys are words, then you could use levenshtein to find all matching words within an edit distance of 1, for example. And that's a pretty easy and well supported thing you can do via the `fst` crate.

**Comment by mre** [comment]
Okay, I see. The `body` field gets sanitized by removing non-alphabetic characters and lowercasing every word. Right now, this is what gets ingested into the bloom filter. I'd like to try fst and a 3-gram of the sanitized body to begin with and see how the package size changes if we go from a 3-gram to a 4-gram for example.

From my prior tests, I found that there are indeed a lot of common prefixes (at least for the sample dataset from my blog). I didn't even consider suffixes, but potentially exploiting that would be really nice.

Seeing that it's not an entirely weird idea, I'll go ahead and give it a shot (or somebody else that might find the time). Will report back here.

**Comment by BurntSushi** [comment]
Great! Would love to hear how it goes and bounce more ideas around.

**Comment by mre** [comment]
Thanks for the quick response and the encouragement. Stay tuned for updates.

**Comment by Jieiku** [comment]
the abridge zola theme supports elasticlunr, tinysearch,and pagefind.

elasticlunr and tinysearch are both whole word searches, but pagefind supports the fuzzy search (shows the results as you type)

elasticlunr: https://abridge.pages.dev/

tinysearch: https://abridge-tinysearch.pages.dev/

pagefind: https://abridge-pagefind.pages.dev/   (if you go here to test it, slowly type the word theme, and see the results update as you type)

I have all 3 searches supported because I see value in all 3 of them. tinysearch is the lightest weight of them all, elasticlunr is simple for people that dont want extra tooling, and pagefind can serve up chunks instead of the entire index, which if you had a massive site like 1,000s of posts then pagefind might pull ahead in terms of data usage, but for smaller blogs Nothing is going to beat how lightweight tinysearch is.

**Comment by mre** [comment]
Just [released](https://github.com/tinysearch/tinysearch/releases/tag/v0.11.0) a version which supports partial word matches.
In my tests, it performed well. :) It's not exactly fuzzy-matching, but it's a start. The binary is not much bigger than before. I've used a very simple data structure for it as described [here](https://github.com/tinysearch/tinysearch/pull/188). It's basically just a concatenated list of unique words per article. Compresses really well with gzip/brotli, so it's just 7% bigger in my tests.

@Jieiku, maybe you can take it for a spin?

**Comment by Jieiku** [comment]
yeah works good (https://abridge-tinysearch.pages.dev/) starts showing results at 3 typed characters.

**Comment by mre** [comment]
oh, thanks for the quick feedback. glad it works.

**Comment by mre** [comment]
I am not really planning to work on this issue anymore, so I will close it as "fixed." (Or at least there is a good enough workaround for my use-case, haha.)

**Comment by epugh** [comment]
Just want to say how much I'm enjoying seeing the vairous updates to this project!

**Comment by mre** [comment]
Yeah, started to have some fun working on the project again. Also added it back to [my website](https://endler.dev/).
